[Transforms] Expand optimizeTan to fold more inverse trig pairs

AZero13 opened the pull request:
optimizeTan has been renamed to optimizeTrigInversionPairs as a result.

Sadly, this is not mathematically true that all inverse pairs fold to x. For example, asin(sin(x)) does not fold to x if x is over 2pi.

AZero13 committed:
[Transforms] Add pre-commit tests

Merge tan-nofastmath.ll and tan.ll into trig.ll

dtcxzyw commented:
Could you please add some regression tests?

AtariDreams commented:
Upd

> Could you please add some regression tests?

Done!

dtcxzyw commented on llvm/test/Transforms/InstCombine/trig-nofastmath.ll:
```diff
@@ -0,0 +1,59 @@
+; RUN: opt < %s -instcombine -S | FileCheck %s
```
Please generate check lines with `llvm/utils/update_test_checks.py`.

dtcxzyw commented on llvm/test/Transforms/InstCombine/trig.ll:
```diff
@@ -0,0 +1,54 @@
+; RUN: opt < %s -instcombine -S | FileCheck %s
+
+define float @mytan(float %x) {
+  %call = call fast float @atanf(float %x)
```
Same problem as https://github.com/llvm/llvm-project/pull/76641#discussion_r1448163247.

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2635,6 +2635,99 @@ Value *LibCallSimplifier::optimizeTan(CallInst *CI, IRBuilderBase &B) {
+  if (!OpC)
+    return Ret;
+
+  // Both calls must be 'fast' in order to remove them.
+  if (!CI->isFast() || !OpC->isFast())
+    return Ret;
```
I'd assume this only needs reassoc + no-errno/readnone callsites

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2635,6 +2635,99 @@ Value *LibCallSimplifier::optimizeTan(CallInst *CI, IRBuilderBase &B) {
+    return Ret;
+
+  // Both calls must be 'fast' in order to remove them.
+  if (!CI->isFast() || !OpC->isFast())
```
Ditto

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2635,6 +2635,99 @@ Value *LibCallSimplifier::optimizeTan(CallInst *CI, IRBuilderBase &B) {
+  return Ret;
+}
+
+Value *LibCallSimplifier::optimizeATanh(CallInst *CI, IRBuilderBase &B) {
```
These functions are all essentially identical. Can you split the function to inverse pair used by a common function?

arsenm commented on llvm/test/Transforms/InstCombine/trig-nofastmath.ll:
```diff
@@ -0,0 +1,59 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+define float @mytan(float %x) {
```
Can you add baseline tests as a pre-commit?

arsenm commented on llvm/test/Transforms/InstCombine/trig-nofastmath.ll:
```diff
@@ -0,0 +1,59 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+define float @mytan(float %x) {
```
Use a better function name, rather than my + second function name

AtariDreams replied:
I was just following what the previous test owner wrote. I'll probably change the original test names too then for consistency

arsenm commented on llvm/test/Transforms/InstCombine/trig-nofastmath.ll:
```diff
@@ -0,0 +1,59 @@
+define float @mytan(float %x) {
+; CHECK-LABEL: define float @mytan(
+; CHECK-SAME: float [[X:%.*]]) {
+; CHECK-NEXT:    [[CALL:%.*]] = call float @atanf(float [[X]])
+; CHECK-NEXT:    [[CALL1:%.*]] = call float @tanf(float [[CALL]])
```
All of these tests don't have flags. You should have all the tests with the necessary flags, and then some negative tests with missing flags

AtariDreams replied:
I think this is what the nofastmath path is for but okay.

AZero13 committed:
[Transforms] Expand optimizeTan to fold more inverse trig pairs

optimizeTan has been renamed to optimizeTrigInversionPairs as a result.

Sadly, this is not mathematically true that all inverse pairs fold to x.

For example, asin(sin(x)) does not fold to x if x is over 2*pi.

AtariDreams commented:
@dtcxzyw @arsenm I think this is ready for review and merge.

AtariDreams commented:
@nikic Thoughts on this?

arsenm commented on llvm/include/llvm/Transforms/Utils/SimplifyLibCalls.h:
```diff
@@ -202,7 +202,7 @@ class LibCallSimplifier {
   Value *optimizeSqrt(CallInst *CI, IRBuilderBase &B);
   Value *optimizeSinCosPi(CallInst *CI, bool IsSin, IRBuilderBase &B);
-  Value *optimizeTan(CallInst *CI, IRBuilderBase &B);
+  Value *optimizeTrigInverionPairs(CallInst *CI, IRBuilderBase &B);
```
Typo, 'Inverion'. Inverse or Inversion?

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2603,13 +2604,27 @@ Value *LibCallSimplifier::optimizeSqrt(CallInst *CI, IRBuilderBase &B) {
-Value *LibCallSimplifier::optimizeTan(CallInst *CI, IRBuilderBase &B) {
+Value *LibCallSimplifier::optimizeTrigInverionPairs(CallInst *CI,
+                                                    IRBuilderBase &B) {
+  static const std::map<llvm::StringRef, LibFunc> trigInverses = {
```
Don't use std::map. Can create an alphabetically sorted constant table and search with lower_bound

jcranmer-intel replied:
`llvm::StringSwitch` is also a workable option.

AtariDreams commented:
@arsenm Can you review this?

AtariDreams commented:
@dtcxzyw Is this ready to merge?

AtariDreams commented:
@arsenm

AtariDreams commented:
@dtcxzyw Ready!

AtariDreams commented:
@arsenm Can we merge this please?
